### DISCUSSION TITLE
how do I place an SVG logo on a photo?

### DISCUSSION THREAD

**Discussion by PontiacCZ** [ImageMagick/ImageMagick#2173] [Help] [answered]
I have a **photo.jpg** and **logo.svg**. Could someone please explain to me how to construct the command line with these two sources if I need to:

**photo.jpg:**
- resize to fit into 2000 x 2000 px square
- normalize
- put text "Annotation" to the right bottom corner (20 pixels from the right and bottom border)

**logo.svg:**
- resize into 100 x 100 px square
- put it into left top corner of the photo.jpg, 20 px from the borders

How the command line would look like? I only have experience with one image but I do not understand how imagemagick works with more image sources. I read the [Anatomy of the Command Line](https://imagemagick.org/script/command-line-processing.php) but I am still confused. I was not able to get the principles from this article.

**Comment by snibgo** [comment]
What version of IM, on what platform?

Do you want to place the resized logo over the original photo.jpg, or photo.jpg resized, normalized and annotated?

**Comment by PontiacCZ** [comment]
Ah, OK.
Windows 10 64bit.
ImageMagick 7.0.10-19 Q16 x64

The logo should be placed over the already processed photo.jpg.

**Comment by snibgo** [comment]
Windows BAT script:
```
magick ^
  photo.jpg ^
  -resize 2000x2000 ^
  -normalize ^
  -gravity SouthEast -annotate 0 "Hello world" ^
  +write photo1.jpg ^
  ( logo.svg ^
    -resize 100x100 ^
  ) ^
  -gravity NorthEast ^
  -compose Over -composite ^
  out.png
```
This makes two outputs: photo1.jpg and out.png. You could write the first to photo.jpg if you want, to overwrite the input file.

**Comment by PontiacCZ** [comment]
Just tried it. It generated two images: photo1.jpg - just the photo resized, normalize, annotated.
out.png - contains the logo as well.

A few things:
- the logo is a solid white rectangle. Maybe add `-background none`?
- its in the top right corner instead of top left but that's just a mistype, I'll change the `NorthEeast` to `NorthWest`
- what about those 20 px offsets?
- why do you use `+write` with the plus symbol instead of minus? I see the [manual about write](https://imagemagick.org/script/command-line-options.php#write) but don't understand what they mean by this "restoring to its original state".

**Comment by snibgo** [comment, marked as answer]
Yes, inserting "-background None" immediately before "logo.svg" might do what you want. It depends on the SVG file.

Sorry, yes, you want top-left, so use "-gravity NorthWest".

Use "-geometry +20+20" before "-composite".

If I used "-write" with a minus, IM might modify the image in memory before writing the output, and would then use that modified image in further processing. "+write" with a plus tells IM to clone the image, and any modifications it makes will be to the clone which gets written, and the clone is removed, so further processing will used the un-modified original.

```
magick ^
  photo.jpg ^
  -resize 2000x2000 ^
  -normalize ^
  -gravity SouthEast -annotate 0 "Hello world" ^
  +write photo1.jpg ^
  ( -background None logo.svg ^
    -resize 100x100 ^
  ) ^
  -gravity NorthWest ^
  -geometry +20+20 ^
  -compose Over -composite ^
  out.png
```

**Comment by PontiacCZ** [comment]
Your script works without problems. 

I implemented it into my batch and modified it (specified colors, added a border) to this form:

```
    %ma% ^
    "%%a" ^
    -normalize ^
    -auto-orient ^
    -resize "%size%x%size%" ^
    -gravity SouthEast ^
    -pointsize "%text_size%" ^
    -fill white ^
    -undercolor #00000015 ^
    -annotate "0x0+20+10" "!line1!\n!line2!" ^
    ( -background none %car_logo% -resize "200x200" ) ^
    -gravity NorthWest ^
    -compose over ^
    -composite ^
    -bordercolor black -border "%border_width%x%border_width%" ^
    "!target_file!"

```

but I get weird error:
`
-gravity was unexpected at this time.
`

It is the second gravity because when I remove it, I get:
`
-compose was unexpected at this time.
`

Don't you know why -gravity is unexpected?

**Comment by snibgo** [comment]
You haven't shown your complete script. I suspect you have the IM command within a shell "for...(...)" loop or "if...(...)" etc. If so, then you need to escape the parentheses in the IM command, like this:
```
  ^( -background None logo.svg ^
     -resize 100x100 ^
  ^) ^
```

**Comment by PontiacCZ** [comment]
Thank you! You were right. I actually only needed to escape the closing parenthesis. The batch language in Win is so poorly designed that it needs things like this... :-( 

And thank you very much for all your help. Now my script does what I wanted it to do. Maybe one day I will understand the peculiar imagemagick syntax and why I need to leave a space after or before parentheses... (that took me so much time to figure this out. I'd love to know why the IM author created such a trap).

**Comment by snibgo** [comment]
True, only the closing parenthesis needs escaping. I think good practice is to escape them both.

Unlike many other languages, no IM tokens outside fx expressions are also separators (which mark token boundaries). Spaces are separators. Characters such as +-(){} are not separators, so a string such as "(-background" is read as a single invalid token, not as two separate tokens "(" and "-background".

**Comment by PontiacCZ** [comment]
I see, thanks for this explanation. :-)
